task(fxa-settings): add BentoMenu component

meandavejustice opened the pull request:
## This pull request

- Implements the global Bento menu into fxa settings

## Issue that this pull request solves

Closes: #4867 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).

## Screenshots (Optional)

This is the current state of the mobile view:
![image](https://user-images.githubusercontent.com/1844554/94509303-ed927580-01e1-11eb-9397-82ef548004f3.png)

This is the desktop view:
![image](https://user-images.githubusercontent.com/1844554/94509342-06029000-01e2-11eb-85de-10848f4ff038.png)

This is what the mobile view looks like on https://monitor.mozilla.com (the floating email field is a bug on the monitor site, (https://github.com/mozilla/blurts-server/issues/1951)
![image](https://user-images.githubusercontent.com/1844554/94509431-3ea26980-01e2-11eb-977a-65e99e790d2c.png)

and monitor desktop:
![image](https://user-images.githubusercontent.com/1844554/94509405-2cc0c680-01e2-11eb-8a8c-20fa3b0c10fc.png)


## Other information (Optional)

Here is a checklist of the tasks I need to complete.

- [x] convert the app icons from the current sprite sheet to svgs. I've requested these and am waiting for a link. (this is more of a code quality bug which can be addressed in the next maintenance sprint if we want)
- [x] Make the bento menu fill the full view on mobile, with the close icon. (I wanted verify that we want to follow this pattern in settings (@pwalm), it seems reasonable 2 me. 
- [x] convert the css rules added in `tailwind.scss` to classes, (this is tied to the svg conversion)
- [x] Add a couple more unit tests.
- [x] Finish adding a11y attributes and get a review from someone who is qualified to do such things.
- [x] pop this into the header/nav (note on this below)

I'm opening this draft pr to show the current state of the bento menu, I'll need to switch over to address #4874 ([FXA-1583](https://jira.mozilla.com/browse/FXA-1583))so that it can make it into the current sprint also, and so that we have a good place for the bento menu to fit into the nav.

meandavejustice commented:
Found and replaced most of the svgs, will finish those up in the morning

LZoog commented on packages/fxa-settings/src/components/BentoMenu/index.tsx:
```diff
@@ -0,0 +1,137 @@
+                  </ul>
+                </div>
+                <LinkExternal
+                  className="link-blue text-xs w-full text-center block mt-2 p-2 hover:bg-grey-100"
```
`m-2` looks a bit closer to Monitor styles than `mt-2`.

LZoog commented on packages/fxa-settings/src/styles/tailwind.scss:
```diff
@@ -37,3 +38,14 @@ body {
+/* stylelint-disable */
+@responsive {
+  .drop-down-menu {
+    @apply w-64 bg-white absolute shadow-md rounded;
```
The Monitor menu is definitely more rounded (maybe `rounded-lg`) but I'd say match whatever Figma is looking like on the avatar menu and we can check with design later.

meandavejustice replied:
updated to rounded-lg, I think it looks okay on both of them now

LZoog commented on packages/fxa-settings/src/styles/tailwind.scss:
```diff
@@ -5,10 +5,11 @@
 @tailwind base;
 @tailwind components;
 @tailwind utilities;
+@tailwind screens;
```
Is this needed for `@responsive` to work? If so maybe that's why it has to be moved in this file.

meandavejustice replied:
It's needed both here and the separate file 👍 that's why we were having such weirdness with the responsive targetting 🤦

LZoog commented on packages/fxa-settings/src/components/BentoMenu/index.tsx:
```diff
@@ -0,0 +1,137 @@
+            title="Firefox Bento Menu"
+            aria-expanded={isRevealed}
+            aria-controls={dropDownId}
+            className="cursor-pointer p-1 pb-0 focus:outline-none transition-standard w-6 mx-6 desktop:mx-8"
```
This should be inside of a button that gets these `aria` values (see `DropDownAvatarMenu`) which will make that `title` show up (and you can take off `role="img"` if it's in a button with a title). 

It's also missing a hover state.

And, this looks better to me at what would be a new class at `w-7` personally, seems a little small despite matching the help icon? But `w-8` seems too large, maybe double check with design?

meandavejustice replied:
Added button, updated attributes, added hover state, and adjusted width, margin, and padding

LZoog commented on packages/fxa-settings/src/components/BentoMenu/index.tsx:
```diff
@@ -0,0 +1,137 @@
+          <div
+            id={dropDownId}
+            data-testid={dropDownId}
+            className="rounded tablet:drop-down-menu tablet:top-10 tablet:-left-54"
```
The caret looks a little off to me at `-left-54` which is `-13.5rem`:

<img width="384" alt="image" src="https://user-images.githubusercontent.com/13018240/94745873-3280ee00-0341-11eb-829a-17ce29282987.png">


`-12.25rem` seems to center it better:

<img width="362" alt="image" src="https://user-images.githubusercontent.com/13018240/94745980-6825d700-0341-11eb-81b3-0cf3a44fb61c.png">

meandavejustice replied:
Cleaned this up and got looked exact

meandavejustice commented on packages/fxa-react/configs/storybooks.js:
```diff
@@ -64,7 +64,7 @@ const customizeWebpackConfig = ({ config, mode }) => {
 
   config.module.rules.push({
-    test: /\.(woff|woff2|eot|ttf)$/,
+    test: /\.(woff|woff2|eot|ttf|png|jpe?g|gif)$/,
```
This was left over from when I initially built it with the sprite sheet from the monitor server, removing.

meandavejustice commented on packages/fxa-settings/src/components/BentoMenu/close.svg:
```diff
@@ -0,0 +1,8 @@
+<svg width="24px" height="24px" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
```
I thought that is what was in https://mozilla.invisionapp.com/d/main#/console/18744572/397559436/preview, but I'm having trouble logging back into invision. It's not sending me the reset link 🤔

meandavejustice replied:
I think it looks better with just the bento menu, so I'm removing the close icon

meandavejustice replied:
I've made that change, we are pulling from `fxa-react/images/close.svg` currently on mobile

meandavejustice commented on packages/fxa-settings/src/components/BentoMenu/index.tsx:
```diff
@@ -0,0 +1,137 @@
+import { ReactComponent as LockwiseIcon } from './lockwise.svg';
+import { ReactComponent as BentoIcon } from './bento.svg';
+import { ReactComponent as CloseIcon } from './close.svg';
+import LinkExternal from 'fxa-react/dist/components/LinkExternal';
```
👍 I think that was a "use quick fix" vs code thing, I'll fix it up

meandavejustice commented on packages/fxa-settings/src/components/BentoMenu/index.tsx:
```diff
@@ -0,0 +1,137 @@
+  const dropDownId = 'drop-down-bento-menu';
+
+  return (
+    <>
```
Copy pasta from dropdown component, removing

meandavejustice commented:
I think I have addressed everything, had some weirdness with the breakpoints which we shouldn't run into in the future since `@tailiwind screens` is now added in the appropriate places.

meandavejustice commented on packages/fxa-react/configs/tailwind.js:
```diff
@@ -57,7 +58,11 @@ module.exports = {
         '1/2': '50%',
         '-px': '-1px',
         '-3': '-0.75rem',
+        '-50': '-11.60rem',
```
`-11.5rem` isn't actually in the center of the button

LZoog reviewed (approved):
Awesome work here, this looks so good across all breakpoints! 🎉
